sdk/core/azure-core/src/main/java/com/azure/core/http/RequestOptions.java

```diff
@@ -0,0 +1,152 @@
+ * other Json building library can be used to achieve similar results.
+ *
+ * {@codesnippet com.azure.core.experimental.http.requestoptions.createjsonrequest}
+ *
+ * Now, this string representation of the JSON request can be set as body of RequestOptions
+ *
+ * {@codesnippet com.azure.core.experimental.http.requestoptions.postrequest}
+ */
+public class RequestOptions {
+    private Consumer<HttpRequest> requestCallback = request -> { };
+    private ResponseStatusOption statusOption = ResponseStatusOption.DEFAULT;
+
+    /**
+     * Gets the request callback, applying all the configurations set on this RequestOptions.
+     * @return the request callback
+     */
+    public Consumer<HttpRequest> getRequestCallback() {
+        return this.requestCallback;
+    }
+
+    /**
+     * Gets under what conditions the operation raises an exception if the underlying response indicates a failure.
+     * @return the configured option
+     */
+    public ResponseStatusOption getStatusOption() {
+        return this.statusOption;
+    }
+
+    /**
+     * Adds a header to the HTTP request.
+     * @param header the header key
+     * @param value the header value
+     *
+     * @return the modified RequestOptions object
+     */
+    public RequestOptions addHeader(String header, String value) {
+        this.requestCallback = this.requestCallback.andThen(request ->
+            request.getHeaders().set(header, value));
+        return this;
+    }
+
+    /**
+     * Adds a query parameter to the request URL.
+     * @param parameterName the name of the query parameter
+     * @param value the value of the query parameter
+     * @return the modified RequestOptions object
+     */
+    public RequestOptions addQueryParam(String parameterName, String value) {
+        this.requestCallback = this.requestCallback.andThen(request -> {
+            String url = request.getUrl().toString();
+            request.setUrl(url + (url.contains("?") ? "&" : "?") + parameterName + "=" + value);
+        });
+        return this;
+    }
+
+    /**
+     * Adds a custom request callback to modify the HTTP request before it's sent by the HttpClient.
+     * The modifications made on a RequestOptions object is applied in order on the request.
+     *
+     * @param requestCallback the request callback
+     * @return the modified RequestOptions object
+     */
+    public RequestOptions addRequestCallback(Consumer<HttpRequest> requestCallback) {
+        this.requestCallback = this.requestCallback.andThen(requestCallback);
+        return this;
+    }
+
+    /**
+     * Sets the body to send as part of the HTTP request.
+     * @param requestBody the request body data
+     * @return the modified RequestOptions object
+     */
+    public RequestOptions setBody(BinaryData requestBody) {
+        this.requestCallback = this.requestCallback.andThen(request -> {
+            request.setBody(requestBody.toBytes());
+        });
```

Comment:
We should not chain this to the callback eagerly. If the user decides to unset the body by calling `setBody(null)` we shouldn't apply 2 consumers - to set and then unset the body.

Reply:
not supporting unsetting body for now